### PR TITLE
Change scopes for absent application/client to restrict to empty set

### PR DESCRIPTION
**Pull request by jasl**
refer PR #577 

The work is not finish, the problem is `valid_scopes` in `/lib/doorkeeper/oauth/helpers/scope_checker.rb` change old behaviour and break 3 cases.

some testing code
```ruby
application = Doorkeeper::Application.first
access_token = Doorkeeper::AccessToken.first
user = User.first

application.scopes # => ['public', 'write']
access_token.scopes # => ['public', 'write', 'yooo']

v = Validator.new('public write', access_token, application)
v.send :@valid_scopes # => ['public', 'write'] MAY NOT MAKE SENSE HERE
v.parsed_scopes # => ['public', 'write']

Doorkeeper::AccessToken.matching_token_for(application, user.id, 'public write yooo') == access_token # => true
```

### CODE REVIEW

**Comment by jasl** [comment]
ping @tute

**Comment by tute** [comment]
Are we doing correctly this check in...?

* `AccessTokenMixin#scopes_match?`
* `PasswordAccessTokenRequest#validate_scopes`
* `ClientCredentialsRequest#validate_scopes`

If we need to change code to handle those checks, we'll need some refactor so we can write this logic only once. This area of the code is very sensitive, and should be as simple as we can.

Also, what happens with applications with empty (default) scopes? Should they allow any scope?

**Comment by tute** [comment]
Let's address most of Hound comments, and make CI green.

File: lib/doorkeeper/models/access_token_mixin.rb [hunk 1]
```diff
@@ -53,18 +53,20 @@ def matching_token_for(application, resource_owner_or_id, scopes)
...
 
-      def scopes_match?(token_scopes, param_scopes, app_scopes)
-        (!token_scopes.present? && !param_scopes.present?) ||
-          Doorkeeper::OAuth::Helpers::ScopeChecker.match?(
-            token_scopes.to_s,
-            param_scopes,
-            app_scopes
-          )
+      def scopes_match?(token, param_scopes, application)
+        return false unless token.present?
```
Comment (tute) [inline lib/doorkeeper/models/access_token_mixin.rb]: Let's use `if` whenever possible: `return false if token.empty?`.

File: lib/doorkeeper/models/access_token_mixin.rb [hunk 2]
```diff
@@ -53,18 +53,20 @@ def matching_token_for(application, resource_owner_or_id, scopes)
...
+        return true if token.scopes.blank? && param_scopes.blank?
```
Comment (tute) [inline lib/doorkeeper/models/access_token_mixin.rb]: Can the `ScopeChecker` handle this case, where both scopes are blank?

File: lib/doorkeeper/oauth/client_credentials/validation.rb [hunk 3]
```diff
@@ -25,18 +25,12 @@ def validate_client
...
 
         def validate_scopes
-          return true unless @request.scopes.present?
-
-          application_scopes = if @client.present?
-                                 @client.application.scopes
-                               else
-                                 ''
-                               end
+          return true unless @request.original_scopes.present?
```
Comment (tute) [inline lib/doorkeeper/oauth/client_credentials/validation.rb]: Let's use `if` in favor of `unless`.

File: lib/doorkeeper/oauth/helpers/scope_checker.rb [hunk 4]
```diff
@@ -18,26 +18,35 @@ def valid?
...
+            valid? && parsed_scopes.equal?(@valid_scopes)
           end
 
           private
 
-          def valid_scopes(server_scopes, application_scopes)
-            if application_scopes.present?
-              server_scopes & application_scopes
+          def valid_scopes(scopes_source, application)
+            # TODO: Refactor this
```
Comment (tute) [inline lib/doorkeeper/oauth/helpers/scope_checker.rb]: Let's address now this `TODO`. A first idea that come to mind is `return` early in positive cases, and leave as implicit return (outside of every conditional, at the end of the function) the `scopes_source.scopes` part.

File: lib/doorkeeper/oauth/scopes.rb [hunk 5]
```diff
@@ -60,6 +60,12 @@ def &(other)
         other_array = other.present? ? other.all : []
         self.class.from_array(all & other_array)
       end
+
+      def equal?(other)
+        self_scopes = self.map(&:to_s).sort
+        other_scopes = other.map(&:to_s).sort
+        (self_scopes.size == other_scopes.size) && ((self_scopes & other_scopes) == self_scopes)
```
Comment (tute) [inline lib/doorkeeper/oauth/scopes.rb]: Let's put a semantic name to the `(self_scopes & other_scopes) == self_scopes`, and extract it as a method.

**Comment by tute** [comment]
Don't forget, @jasl! :) No rush if you have no time, just making sure it's not forgotten.

**Comment by tute** [comment]
Closing due to lack of activity.

**Pull request closed without merging**